<!DOCTYPE html>
<html lang="en-us">
  <head>
    <meta charset="UTF-8">
    <title>Sekator by a-slide</title>
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <link rel="stylesheet" type="text/css" href="stylesheets/normalize.css" media="screen">
    <link href='http://fonts.googleapis.com/css?family=Open+Sans:400,700' rel='stylesheet' type='text/css'>
    <link rel="stylesheet" type="text/css" href="stylesheets/stylesheet.css" media="screen">
    <link rel="stylesheet" type="text/css" href="stylesheets/github-light.css" media="screen">
  </head>
  <body>
    <section class="page-header">
      <h1 class="project-name">Sekator</h1>
      <h2 class="project-tagline">Multithreaded quality and adapter trimmer for PAIRED fastq files (Python2.7/Cython/C)</h2>
      <a href="https://github.com/a-slide/Sekator" class="btn">View on GitHub</a>
      <a href="https://github.com/a-slide/Sekator/zipball/master" class="btn">Download .zip</a>
      <a href="https://github.com/a-slide/Sekator/tarball/master" class="btn">Download .tar.gz</a>
    </section>

    <section class="main-content">
      <h1>
<a id="sekator-02" class="anchor" href="#sekator-02" aria-hidden="true"><span class="octicon octicon-link"></span></a>Sekator 0.2</h1>

<p><strong>Multithreaded quality and adapter trimmer for PAIRED fastq files (Python2.7/Cython/C)</strong></p>

<p><strong>Creation : 2015/03/31</strong></p>

<p><strong>Last update : 2015/03/31</strong> </p>

<h2>
<a id="motivation" class="anchor" href="#motivation" aria-hidden="true"><span class="octicon octicon-link"></span></a>Motivation</h2>

<p>Sekator is a <strong>python2.7, cython 0.21+, C</strong> object oriented semi-compiled program performing fastq quality trimming and adapter trimming</p>

<p>Specific features:</p>

<ul>
<li>The program can parse a pair of fastq files files per sample, but many samples can be analysed together.</li>
<li>Fastq reading and writing are mono-threaded, but the trimming steps are multi-threaded.</li>
<li>The quality trimming step can be performed from both ends of reads with an adjustable sliding windows.</li>
<li>The adapter trimming step is performed by searching imperfect matches of as many adapters as desired (or short sequences) thanks to a fast Smith and Waterman Algorithm coded in C (maintained by <a href="https://github.com/mengyao/Complete-Striped-Smith-Waterman-Library">Mengyao</a>).</li>
</ul>

<h2>
<a id="principle" class="anchor" href="#principle" aria-hidden="true"><span class="octicon octicon-link"></span></a>Principle</h2>

<ol>
<li>A configuration file containing all program parameters (including sample/adpater association) is parsed and thoroughly verified for validity.</li>
<li>Paired fastq paired files are read read by read and sample by sample with a custom Fastq parser supporting <strong>Illumina 1.8 Phred+33 quality encoding only</strong>.</li>
<li>If required, a quality trimming of reads can be performed with a quality sliding windows, starting from both ends of reads. Reads of insufficient quality or too short after trimming are discarded, together with their paired mate.</li>
<li>If required, an adapter trimming of reads can be performed with the adapters provided for each sample. <strong>Imperfect matches can be found anywhere in the reads for as many adapters as required</strong> thanks to an optimized and fast Smith and Waterman Algorithm. If adapters matches are found in a read, the longest part of the read without adapter match is extracted. Reads too short after trimming are discarded, together with their paired mate.</li>
<li>The paired reads that passed thought the trimming steps are subsequently writen in new fastq.gz files (R1 and R2) in Illumina 1.8 Phred+33 quality encoding.</li>
<li>A progress bar indicates the advancement of sequence processing and a report is generated for each sample. </li>
</ol>

<h2>
<a id="dependencies" class="anchor" href="#dependencies" aria-hidden="true"><span class="octicon octicon-link"></span></a>Dependencies</h2>

<p>The program was developed under Linux Mint 17 and was not tested with other OS.
In addition to python2.7 and gcc 4.8 + the following dependencies are required for proper program execution:</p>

<ul>
<li>cython 0.21.2+</li>
<li>python package <a href="http://www.numpy.org/">numpy</a> 1.7.1+</li>
</ul>

<p>If you have pip already installed, enter the following line to install packages: <code>sudo pip install numpy cython</code></p>

<h2>
<a id="get-and-install-sekator" class="anchor" href="#get-and-install-sekator" aria-hidden="true"><span class="octicon octicon-link"></span></a>Get and install Sekator</h2>

<ul>
<li><p>Clone the repository or download the archive <code>git clone https://github.com/a-slide/Sekator.git</code></p></li>
<li><p>Enter the src folder of the program folder</p></li>
<li><p>Compile the C/Cython sources with the Makefile <code>make</code> or the setup.py <code>python setup.py build_ext --inplace</code>. This will create the dynamic library AdapterTrimmer.so required for the adapter trimming step.</p></li>
<li><p>Unnecessary files can be removed with the makefile <code>make clean</code></p></li>
<li><p>Make the main script executable <code>sudo chmod u+x Sekator.py</code></p></li>
<li><p>Finally, add Sekator.py to your PATH</p></li>
</ul>

<h2>
<a id="usage" class="anchor" href="#usage" aria-hidden="true"><span class="octicon octicon-link"></span></a>Usage</h2>

<p>In the folder where fastq files will be created</p>

<p>Usage: Sekator.py -c Conf.txt [-i -h]</p>

<pre><code>Options:
  --version     show program's version number and exit
  -h, --help    show this help message and exit
  -c CONF_FILE  Path to the configuration file [Mandatory]
  -i            Generate an example configuration file and exit [Facultative]
</code></pre>

<p>An example configuration file can be generated by running the program with the option -i
The possible options are extensively described in the configuration file.
The program can be tested from the test folder with the dataset provided and the default configuration file.</p>

<pre><code>cd ./test/result
Sekator.py -i
Sekator.py -c Sekator_conf_file.txt
</code></pre>

<h2>
<a id="authors-and-contact" class="anchor" href="#authors-and-contact" aria-hidden="true"><span class="octicon octicon-link"></span></a>Authors and Contact</h2>

<p>Adrien Leger - 2014</p>

<ul>
<li>
<a href="mailto:adrien.leger@gmail.com">adrien.leger@gmail.com</a> - <a href="mailto:adrien.leger@inserm.fr">adrien.leger@inserm.fr</a> - <a href="mailto:adrien.leger@univ-nantes.fr">adrien.leger@univ-nantes.fr</a>
</li>
<li><a href="https://github.com/a-slide">Github</a></li>
<li><a href="http://www.atlantic-gene-therapies.fr/">Atlantic Gene Therapies - INSERM 1089</a></li>
</ul>

      <footer class="site-footer">
        <span class="site-footer-owner"><a href="https://github.com/a-slide/Sekator">Sekator</a> is maintained by <a href="https://github.com/a-slide">a-slide</a>.</span>

        <span class="site-footer-credits">This page was generated by <a href="https://pages.github.com">GitHub Pages</a> using the <a href="https://github.com/jasonlong/cayman-theme">Cayman theme</a> by <a href="https://twitter.com/jasonlong">Jason Long</a>.</span>
      </footer>

    </section>

  
  </body>
</html>

